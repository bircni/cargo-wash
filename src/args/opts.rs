use std::{fs, path::PathBuf};

use clap::Parser;
use log::warn;

use crate::{data::Project, extensions::PathBufExt, utils};

/// Represents the command line options
#[derive(Parser)]
pub struct Opts {
    /// Path to a directory
    #[clap(long, short)]
    pub path: Option<PathBuf>,
    /// Run the program without making any changes
    #[clap(short, long)]
    pub dry_run: bool,
    /// Exclude specific projects (by name)
    #[clap(long, short = 'e')]
    pub exclude: Vec<String>, // Use a vector to allow multiple exclusions
}

impl Opts {
    pub fn check_args(&self) -> anyhow::Result<(Vec<Project>, bool)> {
        let mut projects: Vec<Project> = vec![];
        let path = super::clean_path(self.path.clone())?;

        if path.is_dir() {
            if path.join("target").is_dir() {
                if let Some(p) = utils::is_cargo_project(&path) {
                    if self.exclude.contains(&p.get_name()?) {
                        return Ok((projects, self.dry_run));
                    }
                    let size = utils::get_folder_size(path.join("target"))?;
                    if size > 0 {
                        projects.push(Project::new(&p.get_name()?, p, size));
                    }
                }
            }

            match fs::read_dir(&path) {
                Ok(entries) => {
                    for entry in entries.flatten() {
                        let path = entry.path();
                        if path.is_dir() {
                            let p = utils::is_cargo_project(&path);

                            if let Some(p) = p {
                                if self.exclude.contains(&p.get_name()?) {
                                    return Ok((projects, self.dry_run));
                                }

                                let size = utils::get_folder_size(path.join("target"))?;
                                if size > 0 {
                                    projects.push(Project::new(&p.get_name()?, p, size));
                                }
                            }
                        }
                    }
                }
                Err(e) => warn!("Error reading directory: {}", e),
            }
        } else {
            anyhow::bail!("The provided path is not a directory.");
        }

        Ok((projects, self.dry_run))
    }
}
