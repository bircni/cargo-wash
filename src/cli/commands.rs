use crate::{
    cli::Commands,
    utils::{self},
};

impl Commands {
    /// Show the output for the command selected
    pub fn show(&self) -> anyhow::Result<()> {
        match self {
            Self::Stats(opts) => {
                let (projects, _) = opts.check_args()?;
                utils::show_stats(&projects);
            }
            Self::Clean(opts) => {
                let (projects, dry_run) = opts.check_args()?;
                utils::run_clean(&projects, dry_run, opts.exclude.as_ref())?;
            }
        }

        Ok(())
    }
}
