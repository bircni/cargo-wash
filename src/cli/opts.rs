use clap::Parser;
use parking_lot::RwLock;
use rayon::iter::{ParallelBridge as _, ParallelIterator as _};
use std::fs;
use std::path::PathBuf;
use std::sync::Arc;

use crate::{data::Project, logic};

/// Represents the command line options
#[derive(Parser, Default, Clone)]
pub struct Opts {
    /// Path to the directory from which to start the search for Rust projects
    /// default is `.`
    #[clap(long, short)]
    pub path: Option<PathBuf>,
    /// Exclude the provided folder from the size calculation and cleaning
    /// default is `None` - you can specify multiple folders separated by `,`
    #[clap(long, short)]
    pub exclude: Option<String>,
}

impl Opts {
    pub fn check_args(&self) -> anyhow::Result<Vec<Project>> {
        let projects: Arc<RwLock<Vec<Project>>> = Arc::new(RwLock::new(vec![]));
        let path = logic::sanitize_path_input(self.path.clone())?;

        if path.is_dir() {
            logic::check_project(&path, self.exclude.as_ref()).map(|p_opt| {
                if let Some(project) = p_opt {
                    projects.write().push(project);
                }
            })?;

            match fs::read_dir(&path) {
                Ok(entries) => {
                    entries.flatten().par_bridge().for_each(|entry| {
                        let pathbuf = entry.path();
                        if pathbuf.is_dir() {
                            match logic::check_project(&pathbuf, self.exclude.as_ref()) {
                                Ok(p_opt) => {
                                    if let Some(project) = p_opt {
                                        projects.write().push(project);
                                    }
                                }
                                Err(error) => log::warn!("Error checking project: {error}"),
                            }
                        }
                    });
                }
                Err(error) => log::warn!("Error reading directory: {error}"),
            }
        } else {
            anyhow::bail!("The provided path is not a directory.");
        }

        Ok(projects.read().to_vec())
    }
}
