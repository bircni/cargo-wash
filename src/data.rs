use core::fmt::{self, Display};
use std::path::{Path, PathBuf};

/// Represents the size unit
/// B = Bytes
/// KB = Kilobytes
/// MB = Megabytes
/// GB = Gigabytes
#[derive(strum_macros::Display, Clone, Copy, PartialEq, PartialOrd, Debug)]
pub enum SizeUnit {
    B,
    GB,
    KB,
    MB,
}

/// Represents a size in bytes with a unit
#[derive(Clone, Copy, PartialEq, PartialOrd, Debug)]
pub struct Size {
    pub unit: SizeUnit,
    pub value: f64,
}

impl Size {
    pub const fn new(value: f64, unit: SizeUnit) -> Self {
        Self { unit, value }
    }

    /// Returns the size in bytes
    #[expect(clippy::cast_possible_truncation, reason = "Ok here")]
    #[expect(clippy::cast_sign_loss, reason = "Ok here")]
    pub fn size_in_bytes(&self) -> u64 {
        let multiplier = match self.unit {
            SizeUnit::B => 1,
            SizeUnit::KB => 1024,
            SizeUnit::MB => 1024 * 1024,
            SizeUnit::GB => 1024 * 1024 * 1024,
        };
        (self.value * f64::from(multiplier)) as u64
    }

    /// Converts a size in bytes to a `Size` struct
    #[expect(clippy::cast_precision_loss, reason = "Ok here")]
    pub fn to_size(bytes: u64) -> Self {
        const KB: u64 = 1024;
        const MB: u64 = KB * 1024;
        const GB: u64 = MB * 1024;

        if bytes >= GB {
            Self::new(bytes as f64 / GB as f64, SizeUnit::GB)
        } else if bytes >= MB {
            Self::new(bytes as f64 / MB as f64, SizeUnit::MB)
        } else if bytes >= KB {
            Self::new(bytes as f64 / KB as f64, SizeUnit::KB)
        } else {
            Self::new(bytes as f64, SizeUnit::B)
        }
    }
}

impl Display for Size {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:.2} {}", self.value, self.unit)
    }
}

/// Represents a Rust project
/// with its name, path, and size
#[derive(Clone, Debug)]
pub struct Project {
    pub name: String,
    pub path: PathBuf,
    pub size: Size,
}

impl Project {
    pub fn new<P: AsRef<Path>>(name: &str, path: P, size: u64) -> Self {
        Self {
            name: name.to_owned(),
            path: path.as_ref().to_path_buf(),
            size: Size::to_size(size),
        }
    }
}
