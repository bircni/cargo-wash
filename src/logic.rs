use std::{
    env, fs,
    path::{Path, PathBuf},
    process::Command,
    sync::Arc,
};

use anyhow::Context as _;
use comfy_table::Table;
use parking_lot::RwLock;
use rayon::iter::{IntoParallelRefIterator as _, ParallelIterator as _};

use crate::{
    data::{Project, Size},
    extensions::PathBufExt as _,
};

pub fn sanitize_path_input(dir: Option<PathBuf>) -> anyhow::Result<PathBuf> {
    let path = dir.unwrap_or_else(|| PathBuf::from("."));
    if path == Path::new("/") || path == Path::new(".") {
        env::current_dir().context("Failed to get current directory")
    } else if path == Path::new("..") {
        env::current_dir()
            .context("Failed to get current directory")?
            .parent()
            .context("Failed to get parent directory")
            .map(Path::to_path_buf)
            .context("Failed to convert parent directory to path")
    } else {
        Ok(path)
    }
}

pub fn total_size_of_projects(projects: &[Project]) -> u64 {
    projects
        .par_iter()
        .map(|project| project.size.size_in_bytes())
        .sum()
}

/// Recursively calculate the size of a folder
pub fn get_folder_size<P: AsRef<Path>>(dir: P) -> anyhow::Result<u64> {
    if !dir.as_ref().exists() {
        return Ok(0);
    }
    let mut total_size = 0;

    for entry_res in fs::read_dir(dir)? {
        let entry = entry_res?;
        let path = entry.path();
        let metadata = fs::metadata(&path)?;

        if metadata.is_dir() {
            total_size += get_folder_size(&path)?;
        } else {
            total_size += metadata.len();
        }
    }

    Ok(total_size)
}

#[expect(clippy::print_stdout, reason = "No other way to show the stats")]
pub fn show_stats(projects: &[Project]) {
    let mut sorted_projects: Vec<Project> = projects.to_vec();
    sorted_projects.sort_by(|a, b| a.name.to_lowercase().cmp(&b.name.to_lowercase()));
    let mut table = Table::new();
    table.set_header(vec!["Project", "Size", "Path"]);

    for project in sorted_projects {
        table.add_row(vec![
            &project.name,
            &project.size.to_string(),
            &project.path.to_string_lossy().to_string(),
        ]);
    }

    table.add_row(vec![
        "Total",
        &Size::to_size(total_size_of_projects(projects)).to_string(),
    ]);
    println!("{table}");
}

pub fn print_status(projects: &[Project], cleaned: &[Project], exclude: Option<&String>) {
    let total_size = total_size_of_projects(cleaned);

    let skipped = exclude.map_or_else(String::new, |skip| format!("\n(Skipped: {skip})"));
    log::info!(
        "Cleaned {} ({} Projects)\nProjects: {}{skipped}",
        Size::to_size(total_size),
        projects.len(),
        cleaned
            .iter()
            .map(|p| p.name.clone())
            .collect::<Vec<String>>()
            .join(", ")
    );
}

pub fn run_clean(projects: &[Project], exclude: Option<&String>) -> anyhow::Result<i32> {
    let cleaned_projects = Arc::new(RwLock::new(vec![]));
    let failed_projects = Arc::new(RwLock::new(vec![]));
    // filter excluded projects
    let mut projects_to_clean = projects.to_vec();

    if let Some(excluded_projects) = exclude {
        log::debug!("Excluding folders: {excluded_projects}");
        excluded_projects
            .split(',')
            .for_each(|ex| projects_to_clean.retain(|project| project.name != ex.trim()));
    } else {
        log::debug!("No folder excluded");
    }

    projects_to_clean.par_iter().for_each(|project| {
        if cfg!(test) {
            log::debug!("Would clean: {:?}", project.name);
            return;
        }
        log::debug!("Running `cargo clean` for project: {:?}", project.name);

        let result = Command::new("cargo")
            .arg("clean")
            .current_dir(&project.path)
            .output();

        match result {
            Ok(output) => {
                if output.status.success() {
                    cleaned_projects.write().push(project.clone());
                } else {
                    failed_projects.write().push(project.clone());
                    log::error!(
                        "Failed to clean {}: {}",
                        project.name,
                        String::from_utf8_lossy(&output.stderr)
                    );
                    log::debug!("inputs were: {projects:?}, {exclude:?}");
                }
            }
            Err(e) => log::error!("Failed to clean {}: {}", project.name, e),
        }
    });
    print_status(projects, &cleaned_projects.read(), exclude);
    if failed_projects.read().is_empty() {
        log::info!("All projects cleaned successfully.");
        Ok(0)
    } else {
        log::warn!(
            "Some projects failed to clean: {}",
            failed_projects.read().len()
        );
        anyhow::bail!(
            "Some projects ({}) failed to clean",
            failed_projects.read().len()
        )
    }
}

pub fn run_rebuild(projects: &[Project], exclude: Option<&String>) -> anyhow::Result<()> {
    let mut rebuilt_projects = vec![];
    let mut failed_projects = vec![];
    // filter excluded projects
    let mut projects_to_rebuild = projects.to_vec();

    if let Some(excluded_projects) = exclude {
        log::debug!("Excluding folders: {excluded_projects}");
        excluded_projects
            .split(',')
            .for_each(|ex| projects_to_rebuild.retain(|project| project.name != ex.trim()));
    } else {
        log::debug!("No folder excluded");
    }

    log::info!(
        "Starting clean & rebuild of {} projects - This might take a while",
        projects_to_rebuild.len()
    );

    projects_to_rebuild.sort_by(|a, b| a.name.to_lowercase().cmp(&b.name.to_lowercase()));

    for project in &projects_to_rebuild {
        log::debug!("Cleaning project: {:?}", project.name);
        let clean_result = Command::new("cargo")
            .arg("clean")
            .current_dir(&project.path)
            .output();
        if let Err(e) = clean_result {
            log::error!("Failed to clean {}: {}", project.name, e);
            failed_projects.push(project.clone());
            continue;
        }

        log::debug!("Running `cargo build` for project: {:?}", project.name);

        let result = Command::new("cargo")
            .arg("build")
            .current_dir(&project.path)
            .output();

        match result {
            Ok(output) if output.status.success() => {
                rebuilt_projects.push(project.clone());
                log::info!("Successfully rebuilt {}", project.name);
            }
            Ok(output) => {
                failed_projects.push(project.clone());
                log::error!(
                    "Failed to rebuild {} with exit code {}",
                    project.name,
                    output.status
                );
            }
            Err(e) => {
                failed_projects.push(project.clone());
                log::error!("Failed to rebuild {}: {e}", project.name);
            }
        }
    }
    log::info!(
        "Rebuilt {} projects successfully, failed to rebuild {} projects.",
        rebuilt_projects.len(),
        failed_projects.len()
    );
    if !failed_projects.is_empty() {
        log::warn!("Some projects failed to rebuild: {}", failed_projects.len());
        anyhow::bail!(
            "Some projects ({}) failed to rebuild",
            failed_projects.len()
        )
    }
    Ok(())
}

pub fn check_project(
    path: &PathBuf,
    exclude_folder: Option<&String>,
) -> anyhow::Result<Option<Project>> {
    if let Some(exclude) = exclude_folder {
        if path.to_string_lossy().contains(exclude) {
            log::debug!("Excluding folder: {exclude}");
            return Ok(None);
        }
        log::debug!("Checking folder: {}", path.to_string_lossy());
    }

    let name = &path.get_name()?;
    let size = get_folder_size(path.join("target"))?;

    if size > 0 {
        return Ok(Some(Project::new(name, path, size)));
    }

    Ok(None)
}
