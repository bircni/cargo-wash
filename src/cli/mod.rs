pub mod opts;

use clap::Parser;
use opts::Options;

use crate::commands::{clean, execute, stats};

/// Represents the available commands
#[derive(Parser)]
#[command(author, version, about)]
pub enum Commands {
    /// Execute `cargo clean` on all projects to remove build artifacts
    Clean(Options),
    /// Print statistics about all Rust projects in the directory
    Stats(Options),
    /// Execute different commands on all rust projects in the directory
    Execute(Options),
}

impl Commands {
    /// Run the command based on the provided options
    pub fn run(&self) -> anyhow::Result<()> {
        match self {
            Self::Stats(opts) => {
                let projects = opts.check_args()?;
                stats::show(&projects);
            }
            Self::Clean(opts) => {
                let projects = opts.check_args()?;
                clean::run(&projects, opts.exclude.as_ref())?;
            }
            Self::Execute(opts) => {
                let projects = opts.check_args()?;
                execute::run(&projects, opts)?;
            }
        }

        Ok(())
    }
}
