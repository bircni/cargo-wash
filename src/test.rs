#![expect(clippy::float_cmp, clippy::similar_names, reason = "Tests")]
use anyhow::Context;
use insta::_macro_support;
use std::fs;
use std::path::Path;
use std::path::PathBuf;
use std::process::Command;
use std::str;

use clap::ColorChoice;

use crate::cli;
use crate::cli::opts::Opts;
use crate::data;
use crate::data::Size;
use crate::data::SizeUnit;
use crate::logic;

#[test]
fn test_logger() {
    super::initialize_logger().unwrap();
}

/// From <https://github.com/EmbarkStudios/cargo-deny/blob/f6e40d8eff6a507977b20588c842c53bc0bfd427/src/cargo-deny/main.rs#L369>
/// Snapshot tests for the CLI commands
#[expect(clippy::panic, reason = "Snapshot failed")]
fn snapshot_test_cli_command(app: clap::Command, cmd_name: &str) {
    let mut app_cmd = app
        .color(ColorChoice::Never)
        .version("0.0.0")
        .long_version("0.0.0");

    let mut buffer = Vec::new();
    app_cmd.write_long_help(&mut buffer).unwrap();
    let help_text = str::from_utf8(&buffer).unwrap();

    let snapshot = _macro_support::SnapshotValue::FileText {
        name: Some(cmd_name.into()),
        content: help_text,
    };

    if _macro_support::assert_snapshot(
        snapshot,
        Path::new(env!("CARGO_MANIFEST_DIR")),
        "cli-cmd",
        module_path!(),
        file!(),
        line!(),
        "help_text",
    )
    .is_err()
    {
        panic!("Snapshot failed");
    }

    for cmd in app_cmd.get_subcommands() {
        if cmd.get_name() == "help" {
            continue;
        }

        snapshot_test_cli_command(cmd.clone(), &format!("{cmd_name}-{}", cmd.get_name()));
    }
}

#[test]
fn test_check_project() {
    let res = logic::check_project(&PathBuf::from("../cargo-wash"), None)
        .unwrap()
        .unwrap();
    assert!(res.size > data::Size::to_size(0));
    assert!(res.name == "cargo-wash");
    assert!(res.path == PathBuf::from("../cargo-wash"));
}

#[test]
fn test_clean_path() {
    logic::sanitize_path_input(Some(PathBuf::from("/"))).unwrap();
    logic::sanitize_path_input(Some(PathBuf::from("."))).unwrap();
    logic::sanitize_path_input(Some(PathBuf::from(".."))).unwrap();
    logic::sanitize_path_input(Some(PathBuf::from("test"))).unwrap();
}

#[test]
fn test_cli_snapshot() {
    use clap::CommandFactory as _;

    insta::with_settings!({
        snapshot_path => "../tests/snapshots",
    }, {
        snapshot_test_cli_command(
            super::Commands::command().name("cargo_wash"),
            "cargo_wash",
        );
    });
}

#[test]
fn test_commands() {
    let opts = Opts::default();
    let command_stats = cli::Commands::Stats(opts);
    command_stats.show().unwrap();
    let opts2 = Opts {
        path: Some(PathBuf::from("/not_existing")),
        ..Default::default()
    };
    assert!(cli::Commands::Stats(opts2.clone()).show().is_err());
    assert!(cli::Commands::Clean(opts2).show().is_err());
    let opts3 = Opts {
        path: Some(PathBuf::from(".")),
        ..Default::default()
    };
    cli::Commands::Clean(opts3).show().unwrap();
}

#[test]
fn test_get_folder_size() {
    let size = logic::get_folder_size("src").unwrap();
    assert!(size > 0);
}

#[test]
fn test_run_clean_excluded() {
    let opts = Opts::default();
    let projects = opts.check_args().unwrap();
    let exclude = "cargo-wash, target".to_owned();
    let result = logic::run_clean(&projects, Some(&exclude));
    assert!(result.is_ok(), "Test failed: {}", result.unwrap_err());
    assert!(
        *result.as_ref().unwrap() == 0,
        "Tests failed: {}",
        result.unwrap_err()
    );
}

#[test]
fn clean_test() {
    // create example project
    let tmp_dir = tempfile::tempdir().unwrap();
    let example_project = tmp_dir.path().join("example_project");

    let result: anyhow::Result<()> = (|| {
        fs::create_dir_all(&example_project)?;
        Command::new("cargo")
            .arg("init")
            .current_dir(&example_project)
            .output()?;

        Command::new("cargo")
            .arg("build")
            .current_dir(&example_project)
            .output()?;

        let opts = Opts {
            path: Some(example_project.clone()),
            ..Default::default()
        };

        let command = cli::Commands::Clean(opts);
        command.show().context("Could not run command")?;
        // utils::run_clean(&projects, false, Some(&exclude))?;
        Ok(())
    })();

    assert!(result.is_ok(), "Test failed: {}", result.unwrap_err());
}

#[test]
fn rebuild_test() {
    // create example project
    let tmp_dir = tempfile::tempdir().unwrap();
    let example_project = tmp_dir.path().join("example_project");

    let result: anyhow::Result<()> = (|| {
        fs::create_dir_all(&example_project)?;
        Command::new("cargo")
            .arg("init")
            .current_dir(&example_project)
            .output()?;

        Command::new("cargo")
            .arg("build")
            .current_dir(&example_project)
            .output()?;

        let opts = Opts {
            path: Some(example_project.clone()),
            ..Default::default()
        };

        let command = cli::Commands::Rebuild(opts);
        command.show().context("Could not run command")?;
        Ok(())
    })();

    assert!(result.is_ok(), "Test failed: {}", result.unwrap_err());
}

#[test]
fn test_size_in_bytes() {
    let size_b = Size::new(500.0, SizeUnit::B);
    assert_eq!(size_b.size_in_bytes(), 500);

    let size_kb = Size::new(1.0, SizeUnit::KB);
    assert_eq!(size_kb.size_in_bytes(), 1024);

    let size_mb = Size::new(1.0, SizeUnit::MB);
    assert_eq!(size_mb.size_in_bytes(), 1024 * 1024);

    let size_gb = Size::new(1.0, SizeUnit::GB);
    assert_eq!(size_gb.size_in_bytes(), 1024 * 1024 * 1024);
}

#[test]
fn test_to_size() {
    let size = Size::to_size(500);
    assert_eq!(size.unit, SizeUnit::B);
    assert_eq!(size.value, 500.0);

    let size = Size::to_size(1024);
    assert_eq!(size.unit, SizeUnit::KB);
    assert_eq!(size.value, 1.0);

    let size = Size::to_size(1024 * 1024);
    assert_eq!(size.unit, SizeUnit::MB);
    assert_eq!(size.value, 1.0);

    let size = Size::to_size(1024 * 1024 * 1024);
    assert_eq!(size.unit, SizeUnit::GB);
    assert_eq!(size.value, 1.0);
}

#[test]
fn test_display() {
    let size = Size::new(1.2345, SizeUnit::MB);
    assert_eq!(format!("{}", size), "1.23 MB");

    let size = Size::new(2048.0, SizeUnit::B);
    assert_eq!(format!("{}", size), "2048.00 B");
}

#[test]
fn test_round_trip_conversion() {
    let original = Size::new(1.5, SizeUnit::MB);
    let bytes = original.size_in_bytes();
    let converted = Size::to_size(bytes);

    assert_eq!(converted.unit, SizeUnit::MB);
    assert!((converted.value - 1.5).abs() < 0.01);
}
