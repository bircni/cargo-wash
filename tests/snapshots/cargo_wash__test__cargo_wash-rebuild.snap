---
source: src/test.rs
expression: help_text
---
Rebuild all Rust projects in the directory

Usage: rebuild [OPTIONS]

Options:
  -p, --path <PATH>
          Path to the directory from which to start the search for Rust projects default is `.`

  -e, --exclude <EXCLUDE>
          Exclude the provided folder from the size calculation and cleaning default is `None` - you can specify multiple folders separated by `,`

  -h, --help
          Print help

  -V, --version
          Print version
